### PR TITLE
Do not include undefined in indexed access on tuple types within range of total fixed elements

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/54420

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16230,6 +16230,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return type.elementFlags.length - findLastIndex(type.elementFlags, f => !(f & flags)) - 1;
     }
 
+    function getTotalFixedElementCount(type: TupleType) {
+        return type.fixedLength + getEndElementCount(type, ElementFlags.Fixed);
```
Comment (jakebailey) [inline src/compiler/checker.ts:16234]: Does every use of `fixedLength` go through this function now? If so, does that imply that this property should be set to this value instead? Or that other uses are also subtly wrong?

Comment (Andarist) [reply src/compiler/checker.ts:16234]: > Does every use of fixedLength go through this function now? If so, does that imply that this property should be set to this value instead?

No, many call sites are interested in the **leading** fixed length so they are not wrong right now and they need the value that is currently held in this property.

> Or that other uses are also subtly wrong?

Some places might still be subtly wrong. I'll likely investigate this if I don't forget but that likely can be just done in followup PRs.

File: tests/baselines/reference/variadicTuples1.types [hunk 2]
```diff
@@ -400,7 +400,7 @@ function f3<T extends unknown[]>(t: [string, ...T, number]) {
 
     let [c1, c2, ...cx] = t;  // string, [string, ...T, number][1], (number | T[number])[]
```
Comment (jakebailey) [inline tests/baselines/reference/variadicTuples1.types:401]: This comment is now wrong (and I think others in this PR too); may want to update them.

**Review by jakebailey — APPROVED** [review]
Just generally, I think this looks fine; the affect on the baselines looks positive to my mind and there's no breaks in the test suites either. Not sure if anyone else should give this a quick pass either, though.

File: tests/baselines/reference/variadicTuples1.types (hunk 2 above)
Comment (Andarist) [reply tests/baselines/reference/variadicTuples1.types]: good catch, I fixed the ones corresponding to the baselines changes

**Pull request merged**